crates/bevy_ecs/hecs/src/dynamic_query.rs

```diff
@@ -0,0 +1,419 @@
+use alloc::sync::Arc;
+use bevy_utils::HashSet;
+use core::{
+    alloc::Layout,
+    any::TypeId,
+    hash::Hash,
+    mem::{align_of, size_of},
+    ptr::{slice_from_raw_parts, slice_from_raw_parts_mut, NonNull},
+};
+use std::{sync::Mutex, vec::Vec};
+
+use crate::{archetype::ComponentIdSet, Access, Archetype, ComponentId, Entity, TypeInfo};
+
+#[derive(Default)]
+pub struct DynamicComponentInfoRegistry {
+    /// Collection of dynamic component info that is kept to ensure the consistency of the info for
+    /// dynamically created components.
+    dynamic_components: Arc<Mutex<HashSet<DynamicComponentInfo>>>,
+}
+
+impl DynamicComponentInfoRegistry {
+    /// Get the component info for the given Rust type
+    pub fn get_rust_component_info<T: 'static>(&self) -> DynamicComponentInfo {
+        DynamicComponentInfo {
+            id: TypeId::of::<T>().into(),
+            layout: Layout::from_size_align(size_of::<T>(), align_of::<T>()).unwrap(),
+            drop: None,
+        }
+    }
+
+    /// Registers a dynamic component and returns [`DynamicComponentInfo`] which can be used to
+    /// spawn dynamic components with an [`EntityBuilder`].
+    ///
+    /// Returns `None` if a component with the provided ID has already been registered.
+    pub fn register_dynamic_component(
+        &self,
+        external_id: u64,
+        layout: Layout,
+        drop: Option<fn(*mut u8)>,
+    ) -> Option<DynamicComponentInfo> {
+        let info = DynamicComponentInfo {
+            id: ComponentId::ExternalId(external_id),
+            layout,
+            drop,
+        };
+
+        if self.dynamic_components.lock().unwrap().insert(info) {
+            Some(info)
+        } else {
+            None
+        }
+    }
+
+    /// Return the component info for a non-Rust dynamic component. If the dynamic component has not
+    /// been registered yet, it will return None.
+    pub fn get_dynamic_component_info(&self, external_id: u64) -> Option<DynamicComponentInfo> {
+        self.dynamic_components
+            .lock()
+            .unwrap()
+            .iter()
+            .find(|&x| x.id == ComponentId::ExternalId(external_id))
+            .map(Clone::clone)
+    }
+}
+
+lazy_static::lazy_static! {
+    /// The global registry of component info
+    pub static ref DYNAMIC_COMPONENT_INFO_REGISTRY: DynamicComponentInfoRegistry =
+        DynamicComponentInfoRegistry::default();
+}
+
+/// A Query that can be constructed at runtime
+#[derive(Default)]
+pub struct DynamicQuery {
+    /// Whether or not the entity should be queried
+    pub entity: bool,
+    // fields private and then checking the values in the constructor.
+    /// The list of accesses to immutable components
+    pub immutable: Vec<DynamicComponentInfo>,
+    /// The list of accesses to mutable components
+    pub mutable: Vec<DynamicComponentInfo>,
```

Comment:
They need to be full info so that you can query the info later.

Reply:
Where do we need to query the info later?

Reply:
This is actually irrelevant if we are doing the checks upon world insertion. This was only necessary if we were doing consistency checks during this registration, so this will go away anyway.